File: 8.string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string str) {
        auto it  = find_if_not(str.begin(), str.end(), ::isspace);
        if (it == str.end()) {
            return 0;
        }
        long long r = 0;
        int sign = 1;
        if (*it == '+') {
            it++;
        } else if (*it == '-') {
            it++;
            sign = -1;
        }
        if (it == str.end() || !isdigit(*it)) {
            return 0;
        }
        while(it != str.end() && isdigit(*it)) {
            r = r * 10 + (*it - '0');
            it++;
            if (sign * r > INT_MAX) {
                return INT_MAX;
            } else if (sign * r < INT_MIN) {
                return INT_MIN;
            }
        }
        return sign * r;
    }
};

File: 8.string-to-integer-atoi.py
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.lstrip()
        if len(str) == 0:
            return 0
        i = 0
        sign = 1
        r = 0
        if str[0] == '+':
            i += 1
        elif str[0] == '-':
            i += 1
            sign = -1
        while i < len(str) and str[i] >= '0' and str[i] <= '9':
            r = r * 10 + ord(str[i]) - ord('0')
            i += 1
            if sign * r > 2 ** 31 - 1:
                return 2147483647
            elif sign * r < - 2 ** 31:
                return -2147483648
        return sign * r